src/allmydata/watchdog/inotify.py

```diff
@@ -0,0 +1,165 @@
+
+from allmydata.util.pollmixin import PollMixin
+from allmydata.util.assertutil import _assert, precondition
+from allmydata.util import log
+from allmydata.util.encodingutil import unicode_from_filepath
+from allmydata.util.fake_inotify import humanReadableMask, \
+    IN_WATCH_MASK, IN_ACCESS, IN_MODIFY, IN_ATTRIB, IN_CLOSE_NOWRITE, IN_CLOSE_WRITE, \
+    IN_OPEN, IN_MOVED_FROM, IN_MOVED_TO, IN_CREATE, IN_DELETE, IN_DELETE_SELF, \
+    IN_MOVE_SELF, IN_UNMOUNT, IN_Q_OVERFLOW, IN_IGNORED, IN_ONLYDIR, IN_DONT_FOLLOW, \
+    IN_MASK_ADD, IN_ISDIR, IN_ONESHOT, IN_CLOSE, IN_MOVED, IN_CHANGED
+[humanReadableMask, \
+    IN_WATCH_MASK, IN_ACCESS, IN_MODIFY, IN_ATTRIB, IN_CLOSE_NOWRITE, IN_CLOSE_WRITE, \
+    IN_OPEN, IN_MOVED_FROM, IN_MOVED_TO, IN_CREATE, IN_DELETE, IN_DELETE_SELF, \
+    IN_MOVE_SELF, IN_UNMOUNT, IN_Q_OVERFLOW, IN_IGNORED, IN_ONLYDIR, IN_DONT_FOLLOW, \
+    IN_MASK_ADD, IN_ISDIR, IN_ONESHOT, IN_CLOSE, IN_MOVED, IN_CHANGED]
+
+
+TRUE  = 0
+FALSE = 1
+
+NOT_STARTED = "NOT_STARTED"
+STARTED     = "STARTED"
+STOPPING    = "STOPPING"
+STOPPED     = "STOPPED"
+
+class INotifyEventHandler(FileSystemEventHandler):
+
+    def __init__(self, path, mask, callbacks, pending_delay):
+        print "init INotifyEventHandler"
+        FileSystemEventHandler.__init__(self)
+        self._path = path
+        self._mask = mask
+        self._callbacks = callbacks
+        self._pending_delay = pending_delay
+        self._pending = set()
+
+    def is_enabled_for_mask(self, event):
+        if isinstance(event, FileSystemMovedEvent) and self._mask & (IN_MOVED_TO | IN_MOVED_FROM):
+            return True
+        if (isinstance(event, FileModifiedEvent) or isinstance(event, DirModifiedEvent)) \
+           and self._mask & (IN_CLOSE_WRITE | IN_CHANGED):
+            return True
+        if (isinstance(event, FileCreatedEvent) or isinstance(event, DirCreatedEvent)) and self._mask & IN_CREATE:
+            return True
+        if (isinstance(event, FileDeletedEvent) or isinstance(event, DirDeletedEvent)) and self._mask & IN_DELETE:
+            return True
+        return False
+
+    def get_event_mask(self, event):
+        if event.is_directory:
+            mask = IN_ISDIR
+        else:
+            mask = 0 # XXX
+        if isinstance(event, FileModifiedEvent) or isinstance(event, DirModifiedEvent):
+            mask = IN_CHANGED
+        if isinstance(event, FileDeletedEvent) or isinstance(event, DirDeletedEvent):
+            mask = IN_DELETE
+        if isinstance(event, FileCreatedEvent) or isinstance(event, DirCreatedEvent):
+            mask = IN_CREATE
+        return mask
+
+    def process(self, event):
+        event_filepath_u = event.src_path.decode(encodingutil.get_filesystem_encoding())
+
+        if event_filepath_u == unicode_from_filepath(self._path):
+            # ignore events for parent directory
+            return
+        #if not self.is_masked(event):
+        #    return
```

Comment:
We need to decide whether to try to make this work, or whether to delete this code and ``is_enabled_for_mask``. I'm inclined to do the former.